x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -1592,42 +1605,82 @@ public boolean test(String profile, InetSocketAddress peerAddress) {
                             + "[xpack.security.http.ssl.key] or [xpack.security.http.ssl.keystore.path] setting"
                     );
                 }
+                if (SSLService.isSSLClientAuthEnabled(sslConfiguration)) {
+                    populateClientCertificate = (channel, threadContext) -> extractClientCertificates(logger, threadContext, channel);
+                } else {
+                    populateClientCertificate = (channel, threadContext) -> {};
+                }
             } else {
                 sslConfiguration = null;
+                populateClientCertificate = (channel, threadContext) -> {};
             }
+            final TriConsumer<BasicHttpRequest, Channel, ThreadContext> populateThreadContext = (httpRequest, channel, threadContext) -> {
+                dispatcherContext.accept(httpRequest, threadContext);
+                populateClientCertificate.accept(channel, threadContext);
+                RemoteHostHeader.process(channel, threadContext);
+            };
+            final AuthenticationService authenticationService = this.authcService.get();
+            final ThreadContext threadContext = this.threadContext.get();
+            final BiConsumer<BasicHttpRequest, ActionListener<Void>> authenticate = (httpRequest, listener) -> {
+                authenticationService.authenticate(httpRequest, listener.map(authentication -> {
+                    if (authentication == null) {
+                        logger.trace("No authentication available for HTTP request [{}]", httpRequest.uri());
+                    } else {
+                        logger.trace("Authenticated HTTP request [{}] as {}", httpRequest.uri(), authentication);
+                    }
+                    return null;
+                }));
+            };
+            HttpServerTransport.Dispatcher authenticationFailedDispatcher = new HttpServerTransport.Dispatcher() {
+                @Override
+                public void dispatchRequest(RestRequest request, RestChannel channel, ThreadContext threadContext) {
+                    dispatcher.dispatchRequest(request, channel, threadContext);
+                }
+
+                @Override
+                public void dispatchBadRequest(RestChannel channel, ThreadContext threadContext, Throwable cause) {
+                    Exception authenticationException = HttpHeadersAuthenticator.extractAuthenticationException(
+                        channel.request().getHttpRequest()
+                    );
+                    if (authenticationException != null) {
+                        SecurityRestFilter.handleAuthenticationFailed(channel, authenticationException);
+                    } else {
+                        dispatcher.dispatchBadRequest(channel, threadContext, cause);
+                    }
+                }
```

Comment:
Unfortunately the dispatcher needs to have the the "bad request" branch overridden, because of the small differences between the 400 errors and authn failure errors.